### ISSUE TITLE
Truthy value not recognized by CFA as a constant condition in loops

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#51199]
# Bug Report

### 🔎 Search Terms

while loop, infinite, cfa, control flow analysis

### 🕗 Version & Regression Information

This is also the 'Nightly' version in the playground: [nightly TS playground](https://www.typescriptlang.org/play?ts=4.9.0-dev.20221017#code/C4TwDgpgBAIglgcwE4EMwAsByB7AJtAXigG8AoKKXFYFALhPIqgGsIR6BnYJOAOwQDcjAL5CKYFEgi9gAfnrxkaLHghDRpUhAAeYbEmBQAZgFdeAY2BxsvKAgjAAytz4IACtXQAKXqoWJUDBx8AEpOF34AbQBdBgoAGwcocxMkKRlgiH8lINUoIl98MWSbLigJYHR8qBihRgB3dDhEqC8UtOlgTJC4pgB6PspsKF9DPnN4k3woJGxsYAByDhY2Rgo4I1aAQnb0rtUAOgk9nrImJilgVNsK9AOpADcIJA4ILxDiimE18s8jkw43l2nUyByoNAOrBAHx+wIyeSIcP2+COkk6xW+3yAA)

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBAIglgcwE4EMwAsByB7AJtAXigG8AoKKXFYFALhPIqgGsIR6BnYJOAOwQDcjAL5CKYFEgi9gAfnrxkaLHghDRpUhAAeYbEmBQAZgFdeAY2BxsvKAgjAAytz4IACtXQAKXqoWJUDBx8AEpOF34AbQBdBgoAGwcocxMkKRlgiH8lINUoIl98MWSbLigJYHR8qBihRgB3dDhEqC8UtOlgTJC4pgB6PspsKF9DPnN4k3woJGxsYAByDhY2Rgo4I1aAQnb0rtUAOgk9nrImJilgVNsK9AOpADcIJA4ILxDiimE18s8jkw43l2nUyByoNAOrBAHx+wIyeSIcP2+COkk6xW+3yAA)

### 💻 Code

```ts
type DigraphNode = {
  data: {
    key: string;
  };
  parent?: DigraphNode;
};

export function getStringPath(node: DigraphNode): string[] { // Function lacks ending return statement and return type does not include 'undefined'.(2366)
  let currentNode: DigraphNode = node;
  const path = [];

  while (currentNode) {
    // do not include root's key
    if (!currentNode.parent) {
      return path.reverse();
    }
    path.push(currentNode.data.key);
    currentNode = currentNode.parent;
  }
}
```

### 🙁 Actual behavior

I get an error:
> Function lacks ending return statement and return type does not include 'undefined'.(2366)

### 🙂 Expected behavior

I shouldn't get an error because `currentNode` is always truthy, this should work in the same way as `while (true)` ([TS playground](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBAIglgcwE4EMwAsByB7AJtAXigG8AoKKXFYFALhPIqgGsIR6BnYJOAOwQDcjAL5CKYFEgi9gAfnrxkaLHghDRpUhAAeYbEmBQAZgFdeAY2BxsvKAgjAAytz4IACtXQAKXqoWJUDBx8AEpOF34AbQBdBgoAGwcocxMkKRlgiH8lINUoIl98MWSbLigJYHR8qBihRgB3dDhEqC9uEwgQuKYAeh7KbChfQz5zeJN8KCRsbGAAcg4WNkYKOCNWgEIUtOlgTIA6CXTgLrImJilgVNsK9H2pADcIJA4ILxDiimEV8s9Dkw43m2xwOVBo+1YIA+P2Bu0y1VhGVUh0ku2K32+QA)).

**Comment by RyanCavanaugh** [comment]
Interesting case. Right now this check is *almost* entirely syntactic; it does consider exhaustive `switch`es but otherwise doesn't use CFA types at all.

This is probably a Won't Fix unless the fix is extremely cheap, since the logic of "this is the same as a `while (true)`" strongly implies that the program should just be written that way.

**Comment by fatcerberus** [comment]
Wouldn’t it just be the same check as the one that produces “this condition will always be true because…” diagnostics for `if` statements?

**Comment by Andarist** [comment]
> This is probably a Won't Fix unless the fix is extremely cheap, since the logic of "this is the same as a while (true)" strongly implies that the program should just be written that way.

I agree that this is how it should be written and I've rewritten it this way before opening this issue. However, I think that the compiler shouldn't rely on this - that's why I've created an issue. A diagnostic like the one mentioned by @fatcerberus would also be nice but not necessary.

> Wouldn’t it just be the same check as the one that produces “this condition will always be true because…” diagnostics for if statements?

Interestingly, this particular situation doesn't produce such a diagnostic even with the `if` statement and it suffers from the same issue related to the return type:
```ts
type DigraphNode = {
  data: {
    key: string;
  };
  parent?: DigraphNode;
};

export function getStringPath(node: DigraphNode): string[] { // Function lacks ending return statement and return type does not include 'undefined'.(2366)

  let currentNode: DigraphNode = node;
  const path = [];

  if (currentNode) {
    path.push(currentNode.data.key);
    return path.reverse();
  }
}
```
